common/src/main/java/com/wynntils/models/raid/RaidModel.java

```diff
@@ -0,0 +1,209 @@
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.regex.Pattern;
+import net.minecraft.network.chat.Component;
+import net.minecraft.world.entity.decoration.ArmorStand;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class RaidModel extends Model {
+    // For raid entry use title.
+    // For start of a challenge, look for an instructions label.
+    // For end of a challenge look for BLACKSMITH_LABEL label.
+    // For boss intermission look for BOSS_FIGHT_LABEL label.
+    // For boss fight look for the boss' label.
+    // For raid completion/failure use RAID_COMPLETE or RAID_FAILED title.
+    private static final StyledText BLACKSMITH_LABEL = StyledText.fromString("§dBlacksmith");
+    private static final StyledText BOSS_FIGHT_LABEL = StyledText.fromString("§4§l[§cBoss Fight§4§l]");
+    // Title gradually builds up to full "§a§lRAID COMPLETED!"
+    private static final StyledText RAID_COMPLETE = StyledText.fromString("§a§lD C");
+    private static final StyledText RAID_FAILED = StyledText.fromString("§4Raid Failed!");
+
+    @Persisted
+    public final Storage<Map<String, Integer>> bestTimes = new Storage<>(new TreeMap<>());
+
+    private final Map<RaidRoomType, Integer> roomTimers = new EnumMap<>(RaidRoomType.class);
+
+    private long raidStartTime;
+    private long roomStartTime;
+    private Raid currentRaid = null;
+    private RaidRoomType currentRoom = null;
+
+    public RaidModel() {
+        super(List.of());
+    }
+
+    @SubscribeEvent
+    public void onTitle(TitleSetTextEvent event) {
+        Component component = event.getComponent();
+        StyledText styledText = StyledText.fromComponent(component);
+
+        if (currentRaid == null) {
+            currentRaid = Raid.fromTitle(styledText);
+
+            if (currentRaid != null) {
+                // In a raid, set to intro room and start timer
+                currentRoom = RaidRoomType.INTRO;
+                raidStartTime = System.currentTimeMillis();
+            }
+        } else {
+            if (styledText.equals(RAID_COMPLETE)) {
+                int timeTaken = (int) ((System.currentTimeMillis() - raidStartTime) / 1000);
+                // Raid has been completed, post event with time taken in seconds
+                WynntilsMod.postEvent(new RaidFinishedEvent.Completed(currentRaid, timeTaken));
+
+                checkForNewPersonalBest(currentRaid, timeTaken);
+            } else if (styledText.equals(RAID_FAILED)) {
+                // Raid failed, post event with time elapsed in seconds
+                WynntilsMod.postEvent(new RaidFinishedEvent.Failed(
+                        currentRaid, (int) ((System.currentTimeMillis() - raidStartTime) / 1000)));
+            } else {
+                return;
+            }
+
+            currentRaid = null;
+            currentRoom = null;
+            roomTimers.clear();
+        }
```

Comment:
```suggestion
        if (currentRaid == null) {
            currentRaid = Raid.fromTitle(styledText);

            if (currentRaid != null) {
                // In a raid, set to intro room and start timer
                currentRoom = RaidRoomType.INTRO;
                raidStartTime = System.currentTimeMillis();
            }
            return;
        } 
        
        if (styledText.equals(RAID_COMPLETE)) {
            int timeTaken = (int) ((System.currentTimeMillis() - raidStartTime) / 1000);
            // Raid has been completed, post event with time taken in seconds
            WynntilsMod.postEvent(new RaidFinishedEvent.Completed(currentRaid, timeTaken));

            checkForNewPersonalBest(currentRaid, timeTaken);
            
            currentRaid = null;
            currentRoom = null;
            roomTimers.clear();
        } else if (styledText.equals(RAID_FAILED)) {
            // Raid failed, post event with time elapsed in seconds
            WynntilsMod.postEvent(new RaidFinishedEvent.Failed(
                    currentRaid, (int) ((System.currentTimeMillis() - raidStartTime) / 1000)));
          currentRaid = null;
          currentRoom = null;
          roomTimers.clear();
        }
```

else -> return is a bit odd, so maybe 3 lines of code duplication is better. Alternatively, set a boolean and return if it was not either case.